android/src/main/java/com/internxt/mobilesdk/core/Upload.kt

```
@@ -0,0 +1,205 @@
+    val startUploadAdapter = JsonUtils.moshi.adapter(StartUploadResponse::class.java)
+    val startUploadPayloadAdapter = JsonUtils.moshi.adapter(StartUploadPayload::class.java)
+    val base = MobileSdkConfigLoader.getConfig(MobileSdkConfigKey.BRIDGE_URL)
+    val url = "${base}/v2/buckets/${bucketId}/files/start?multiparts=${parts}"
+    val payload = startUploadPayloadAdapter.toJson(startUploadPayload)
+    Logger.info("Starting upload with payload $payload")
+    val request = Request.Builder().header("Authorization", getCredentials()).post(payload.toRequestBody()).url(url).build()
+    val response = HttpClient.fetchSync(request)
+    return startUploadAdapter.fromJson(response.body!!.source())
+  }
+
+  private fun finishNetworkUpload(bucketId: String, finishUploadPayload: FinishUploadPayload): FinishUploadResponse? {
+    val finishUploadAdapter = JsonUtils.moshi.adapter(FinishUploadResponse::class.java)
+    val uploadFailedAdapter = JsonUtils.moshi.adapter(UploadFailedResponse::class.java)
+    val finishUploadPayloadAdapter = JsonUtils.moshi.adapter(FinishUploadPayload::class.java)
+    val base = MobileSdkConfigLoader.getConfig(MobileSdkConfigKey.BRIDGE_URL)
+    val url = "${base}/v2/buckets/${bucketId}/files/finish"
+
+    val payload = finishUploadPayloadAdapter.toJson(finishUploadPayload)
+    Logger.info("Finishing upload with payload $payload")
+    val request = Request.Builder().header("Authorization", getCredentials()).post(payload.toRequestBody()).url(url).build()
+    val response = HttpClient.fetchSync(request)
+
+    if(!response.isSuccessful) {
+      val body = uploadFailedAdapter.fromJson(response.body!!.source())
+
+      Logger.error("Failed upload finish ${body!!.error}")
+      if(body?.error?.startsWith("E11000 duplicate key error") == true) {
+        throw DuplicatedUpload(body.error)
+      }
+
+      throw ApiResponseException("Finish upload not successful")
```

Could we throw the error with the body error message inside? Then, we do not miss the error

---

Still, if the error is not the duplicated upload, we need the response from the server as more context. So what I was referring to was: 

`throw ApiResponseException("Error uploading" + body.error)`

Or something like that